npm/networkPolicyController.go

```diff
@@ -0,0 +1,491 @@
+	// (TODO): where do we use this?
+	timer := metrics.StartNewTimer()
+
+	var err error
+	netpolKey, err := cache.MetaNamespaceKeyFunc(netPolObj)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: while running MetaNamespaceKeyFunc err: %s", err)
+	}
+
+	// Start reconciling loop to eventually meet cached states against the desired states from network policy.
+	// #1 If a new network policy is created, the network policy is not in RawNPMap.
+	// start translating policy and install translated ipset and iptables rules into kernel
+	// #2 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the same object (same UID),
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// #3 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the different object (different UID) due to missing some events for the old object
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// To deal with all three cases, we first delete network policy if possible, then install translated rules into kernel.
+	// (TODO): can optimize logic more to reduce computations. For example, apply only difference if possible like podController
+
+	// Do not need to clean up default Azure NPM chain in deleteNetworkPolicy function, if network policy object is applied soon.
+	// So, avoid extra overhead to install default Azure NPM chain in initializeDefaultAzureNpmChain function.
+	// To achieve it, use flag unSafeToCleanUpAzureNpmChain to indicate that the default Azure NPM chain cannot be deleted.
+	// delete existing network policy
+	err = c.deleteNetworkPolicy(netPolObj, unSafeToCleanUpAzureNpmChain)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: failed to deleteNetworkPolicy due to %s", err)
+	}
+
+	// Install this default rules for kube-system and azure-npm chains if they are not initilized.
+	// Execute initializeDefaultAzureNpmChain function first before actually starting processing network policy object.
+	if err = c.initializeDefaultAzureNpmChain(); err != nil {
+		return fmt.Errorf("[syncNetPol] Error: due to %v", err)
+	}
+
+	ipsMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+	iptMgr := c.npMgr.NsMap[util.KubeAllNamespacesFlag].iptMgr
+	sets, namedPorts, lists, ingressIPCidrs, egressIPCidrs, iptEntries := translatePolicy(netPolObj)
+	for _, set := range sets {
+		log.Logf("Creating set: %v, hashedSet: %v", set, util.GetHashedName(set))
+		if err = ipsMgr.CreateSet(set, append([]string{util.IpsetNetHashFlag})); err != nil {
+			return fmt.Errorf("[syncAddAndUpdateNetPol] Error: creating ipset %s with err: %v", set, err)
+		}
+	}
+	for _, set := range namedPorts {
+		log.Logf("Creating set: %v, hashedSet: %v", set, util.GetHashedName(set))
```

Comment:
Also one more test case with NamedPort network policy, there should be an example in npm/testpolicies/ folder.

Reply:
Do you mean we need tp add another yaml file in npm/testpolicies folder?

Reply:
either use the existing yaml to have a named port UT or for consistency purposes you can create a namedPort policy with createPolicy

Reply:
Good points. applied to test different type of ports.